Changes from fork

ollipa opened the pull request:
This PR contains all the commits from the fork that we maintained before Github project was transferred to flexmock organization.

Closes #34

@bkabrda, I removed RPM specfile and packit conf because I was not sure if there is still a need to keep those files. Let me know if you wish to keep the files and I'll re-add them.

**Breaking changes**:
- Dropped support for Python 2.7, 3.4, 3.5.
- Removed unittest2 integration. unittest2 is not maintained anymore.
- Flexmock needs to be imported explicitly using `from flexmock import flexmock`. The hack that allowed importing flexmock directly using `import flexmock`, doesn't really work well linters and automatic completion. However, I wonder if we should still keep the hack for backwards compatibility.

ollipa committed:
Add Poetry and cleanup repository

ollipa committed:
Refactor tests

ollipa committed:
Refactor flexmock.py

ollipa committed:
Organize code

ollipa committed:
Add type annotations

christophe-riolo committed:
Update packaging

christophe-riolo committed:
Rename into flexmock-plus

ollipa committed:
Update authors

ollipa committed:
Add removed integrations back

ollipa committed:
Remove Pipfile

christophe-riolo committed:
Fix documentation generation

christophe-riolo committed:
Add `CONTRIBUTING.md` file

christophe-riolo committed:
Use context managers to suppress import errors in integrations

christophe-riolo committed:
Split _patch_test_result

The function was quite long and convoluted, as it was patching two methods, we
are rewriting it into two smaller functions.

christophe-riolo committed:
Move logic out of __init__.py

Ideally the __init__.py file should not have any logic. We can very well hook the
integrations straight in the `integrations` module for the same effect.

christophe-riolo committed:
Remove unnecessary Literal

The arguments and the functions signature do not depend on us, it's imposed by
the runner. Since we don't have any leverage on it, we might as well drop a dependency.

ollipa committed:
Fix no re.Pattern in Python 3.6

ollipa committed:
Ignore mypy errors in integrations

ollipa committed:
Run unittest test in Makefile

ollipa committed:
Add tox for running tests with different Python version

ollipa committed:
Run tests and linters in Github Actions

ollipa committed:
Extend contribution documentation

ollipa committed:
Update command to run unittest tests

ollipa committed:
Change CHANGELOG back to rst format

ollipa committed:
Add readthedocs.yml

ollipa committed:
Update project name

ollipa committed:
Update documentation

ollipa committed:
Update Read The Docs build Python version

ollipa committed:
Prepare 0.11.0 release

lachmanfrantisek commented:
>  I removed RPM specfile and packit conf because I was not sure if there is still a need to keep those files. Let me know if you wish to keep the files and I'll re-add them.

This is for the [packit-service](https://github.com/apps/packit-as-a-service) to ensure the package can be built/instaled/tested in RPM-based systems. (Spoiler: I am one of the developers...;)  It can also provide Copr repositories for a branch or releases as well as helping `flexmock` maintainer(s) in Fedora to get the new versions there. I don't want to push you to use it, but let me know if you want to get more from it.

Regarding the specfile, it [can be downloaded](https://packit.dev/docs/faq/#how-can-i-download-rpm-spec-file-if-it-is-not-part-of-upstream-repository) during the process if you do not want to store it in the git repository. I can send a PR if you want.

Nevertheless, thanks for taking care of `flexmock`, we use it a lot in our project.

ollipa commented:
@lachmanfrantisek, in that case I'll put the files back :slightly_smiling_face: I was just not sure if they were still needed.

ollipa commented:
/packit build

lachmanfrantisek commented:
/packit build

note: I've approved the new project location in the meantime.

lachmanfrantisek commented:
> @lachmanfrantisek, in that case I'll put the files back slightly_smiling_face I was just not sure if they were still needed.

Thanks..;) Let me know if you have any trouble with the service.

christophe-riolo commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,63 @@
+
+      - name: Install dependencies
+        run: |
+          poetry config virtualenvs.in-project true
```
what's that line?

ollipa replied:
With this config poetry creates virtual environment in the current folder. Otherwise it would go to the user's home directory.

christophe-riolo commented on CHANGELOG.rst:
```diff
@@ -1,3 +1,57 @@
+Removed
+#######
+
+- Drop Python 2.7, 3.4, 3.5 support.
```
I'm still not sure we should drop 3.4 and 3.5 already

ollipa replied:
Why do you think so? Python 3.4 reached end of life already two years ago and Python 3.5 almost one year. Even pytest has dropped support for Python 3.4. If someone still needs Python 3.4 or 3.5, they can keep using an older flexmock version.

If we keep supporting Python 3.4 & 3.5 we would need to remove type hints.

ollipa commented:
I think the spec file needs some tweaking and it seems that need to bring back setup.py as well. I'll continue tomorrow and let you know if I need help.

christophe-riolo reviewed (commented):
I need to test a couple of things but I somehow cannot install python versions with pyenv, I need to figure that out

ollipa committed:
Update setup.py

ollipa committed:
Update old links

ollipa committed:
Update RPM specfile

ollipa commented:
/packit build

ollipa committed:
Add twisted to test runs

ollipa committed:
Remove nose support

Nose is not maintained anymore.
It might still work with flexmock but
we don't support it officially anymore.

ollipa commented:
@lachmanfrantisek, it seems that the new project location is not in the allowlist anymore for some reason.

mfocko commented:
> it seems that the new project location is not in the allowlist anymore for some reason.

It should be fine now :) (it was missing `.git` suffix for repo-only, in case you wanted to include more repositories, don't hesitate to reach out)

could you please trigger it to check? (I bypass the namespace check as admin)

ollipa commented:
/packit build

ollipa commented:
@lachmanfrantisek, packit builds are successful now. I couldn't get the tests working because flexmock package is not installed in the runner when tests are executed. Is it possible to install the created package before executing tests? The old test script added current folder to PYTHONPATH so it wasn't actually testing the package itself.

TomasTomecek commented:
@ollipa I'm gonna take a look and help you with tests. Will try to make those work in my fork now and let you know.

Edit: Oh, now I understand the ask, it's about running tests during the build. That workflow is deprecated to some extent - will try to port it to the flow you described: install the built RPM and run tests on it.

TomasTomecek committed:
rpm: set %release to 1 after bumping version to 0.11.0

Signed-off-by: Tomas Tomecek <ttomecek@redhat.com>

TomasTomecek committed:
replace %check testing with packit/testing-farm

Tests will be now executed in a dedicated, freshly provisioned VMs

for more info see:
http://tmt.readthedocs.io/
https://packit.dev/testing-farm/

Signed-off-by: Tomas Tomecek <ttomecek@redhat.com>

TomasTomecek commented:
You can see in my PR that tests and linters are now running in two separate steps for the same set of targets: https://github.com/TomasTomecek/flexmock/pull/1

The failures are coming from mypy, not sure if they are false or true positives.

Let me know what you think. You're welcome to cherry-pick those comments and include them here - I can file a PR if you want me to.

ollipa committed:
Do not run linters in Fedora tests

ollipa commented:
@TomasTomecek, thanks! I cherry-picked your commits. Mypy probably failed because either Mypy or Python version didn't match the one we have in our CI. I think it's better not to run linters in Fedora test farm since we already run them in our CI and type checking and linting doesn't affect the runtime behavior.

Is it be possible to move some packit related files to their own subfolder? Ideally it would be nice to have most of the files in their own folder with a README file that would explain what the files do and how they can be maintained.

TomasTomecek commented:
@ollipa except for .packit.yaml, everything can be placed in a subdirectory - I'll try to do that today - will build those changes on top of the current PR HEAD.

TomasTomecek committed:
move Fedora packaging and CI to fedora/ folder

Signed-off-by: Tomas Tomecek <ttomecek@redhat.com>

TomasTomecek commented:
@ollipa I have finally finished the move: https://github.com/TomasTomecek/flexmock/pull/3

please let me know if I should include more documentation how to work with the files/services; sadly two paths cannot be moved there: .fmf/version (requirement from the testing framework) and .packit.yaml (requirement from the CI system)

christophe-riolo commented on pyproject.toml:
```diff
@@ -0,0 +1,102 @@
+
+[tool.pylint.messages_control]
+disable = [
+    "too-many-instance-attributes",
+    "too-few-public-methods",
+    "logging-fstring-interpolation",
+    "too-many-function-args",
+    "too-many-arguments",
+    "bad-continuation",
+    "too-many-locals",
+    "too-many-branches",
+    "too-many-statements",
+    "too-many-public-methods",
+    "protected-access",
+    "broad-except",
+    "fixme",
```
We could try to reduce those at some point, and use local disables, sparingly

ollipa replied:
Yeah lets do some cleanup later. There were quite many places where pylint was complaining so I just disabled some rules globally. I will make an issue about this.

ollipa replied:
https://github.com/flexmock/flexmock/issues/47

christophe-riolo reviewed (approved):
Back from vacation, sorry for the delay in review; this looks good and we can make that release I think !

ollipa commented:
@TomasTomecek, thanks I included your commit. Documentation looks good to me. We can always extend it later if needed.

@adarshk7, do you still want to take a look at the changes before I merge?

adarshk7 reviewed (approved):
Looks good to me :rocket:
